protocols/autonat/src/v2/client/handler/dial_back.rs

```diff
@@ -0,0 +1,142 @@
+use std::{
+    io,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{channel::oneshot, AsyncWriteExt};
+use futures_bounded::StreamSet;
+use libp2p_core::upgrade::{DeniedUpgrade, ReadyUpgrade};
+use libp2p_swarm::{
+    handler::{ConnectionEvent, FullyNegotiatedInbound, ListenUpgradeError},
+    ConnectionHandler, ConnectionHandlerEvent, StreamProtocol, SubstreamProtocol,
+};
+use void::Void;
+
+use crate::v2::{protocol, Nonce, DIAL_BACK_PROTOCOL};
+
+pub struct Handler {
+    inbound: StreamSet<io::Result<IncomingNonce>>,
+}
+
+impl Handler {
+    pub(crate) fn new() -> Self {
+        Self {
+            inbound: StreamSet::new(Duration::from_secs(5), 2),
+        }
+    }
+}
+
+impl ConnectionHandler for Handler {
+    type FromBehaviour = Void;
+    type ToBehaviour = IncomingNonce;
+    type InboundProtocol = ReadyUpgrade<StreamProtocol>;
+    type OutboundProtocol = DeniedUpgrade;
+    type InboundOpenInfo = ();
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(ReadyUpgrade::new(DIAL_BACK_PROTOCOL), ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        match self.inbound.poll_next_unpin(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(None) => Poll::Pending,
```

Comment:
This is incorrect, you can only return `Poll::Pending` when you've either registered a waker or something you `poll`ed returned `Poll::Pending`.

I'd suggest to `loop` here and `continue` in all branches that didn't return `Poll::Pending`.

Reply:
yes that's good. I think I misunderstood something here.